### PR TITLE
New Module: Homectl module for managing systemd-homed

### PR DESCRIPTION
**Pull request by jameslivulpi**
##### SUMMARY

#3931
New module (WIP) for adding module homectl to manage systemd-homed user accounts.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME

homectl

##### ADDITIONAL INFORMATION
WIP for now creates user account and removes them.

Would appreciate any feedback or issues anyone sees / suggestions/ comments

### CODE REVIEW

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:304]: If the module does no idempotency checks except existence, this needs to be clearly documented.

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've added a first few comments.

**Comment by jameslivulpi** [comment]
Felix thanks so much on comments. I will address them all at some point today!

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:32]: Trust me I am just as confused, from user record [doc](https://systemd.io/USER_RECORD/) it mentions:
```
password → an array of strings, each containing a plain text password.
```

Perhaps this is a typo, I will run more tests and find out.

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:45]: So when creating a user, if storage mechanism is not given, homed will automatically use 'classic' perhaps the default is not needed within the module? homectl doc mentions: `If this is not set classic is the implied default.`

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:151]: Yes, I agree. This is only making sure the service is in 'active' state and let the user know if not. I think this information is good for the user to receive to point them in right direction that they should fix this.

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:305]: Can you elaborate? When a user is to be added or removed this check will happen. I plan to add more code to modify a current user which will also check of user exists and grabs the current metadata for that user.

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:30]: Agreed, I am just concerned about plaintext password being sent for man in middle attacks.

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py]: removed the ` array of strings` line. fixed.

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:316]: I think I see what you are saying.

So here:

```
else:
     result['msg'] = "User already Exists!"
```

should also make sure the "chaged" would be False such as:

```python3
else:
      result['changed'] = False
      result['msg'] = "User already Exists!"
```

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:19]:
```suggestion
    - Manages a user's home directory managed by systemd-homed.
```
Sentences in descriptions should always be complete. (The only exception is `short_description`.)

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:33]:
```suggestion
            - This is required for I(state=present) and I(state=modify).
```

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:50]:
```suggestion
            - Human readable value such as C(10G), C(10M), or C(10B).
```

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:84]:
```suggestion
            - This is not where the user's data is actually stored, see I(imagepath) for that.
```

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:93]:
```suggestion
            - Sets the UID of the user.
```
(Same below.)

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:94]:
```suggestion
            - If using I(gid) homed requires the value to be the same.
```
(Same below.)

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:115]:
```suggestion
            - If not specified homed by default uses C(/bin/bash).
```

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:167]:
```suggestion
    description: A json dictionary returned from C(homectl inspect -j).
```

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:436]:
```suggestion
            result['msg'] = "User does not exist!"
```

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:38]: Modify is **not** a state. A state is "present" or "absent".

Modification should always be part of "present".

File: plugins/modules/system/homectl.py
Comment (felixfontein) [reply plugins/modules/system/homectl.py:455]: Yes, it should set `changed=False`, **but** the module documentation should also say that no changes are made when `state=present`. This is different from most other Ansible modules. (Your `modify` state should be a part of `present`.)

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:38]: I see, I will take care of this.

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:127]:
```suggestion
            - If not specified homed by default uses C(/etc/skel).
```
Only ever use double backticks in documentation if `C(...)` cannot be used (because the text to quote contains `(` or `)`). Same below.

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:185]: How about also using some female names in the example?
```suggestion
- name: Add the user 'alice' with a zsh shell, uid of 1000, and gid of 2000
  community.general.homectl:
    name: alice
```

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:319]: What is `use_unsafe_shell=True` for here and in the other calls? `use_unsafe_shell=True` should only be used if absolutely necessary.

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:375]: Are this and the other fields always present? If not necessarily, use
```suggestion
            user_metadata.pop('signature', None)
```
instead.

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:384]: Is there a reason you sometimes use double quotes and sometimes single quotes for strings?

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:341]: 1. There is a space missing after the `return` keyword.
2. The brakets are superfluous.
```suggestion
        return self.module.run_command(cmd, use_unsafe_shell=True)
```
Same for similar calls.

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:604]: This looks wrong. At least the change detection should be run before determining that `changed=True`.

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:354]: I don't think this should be run except if a change is to be expected. (Also the resize operation does not seem to count for changed?)

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:127]: Thanks, sorry I am still getting used to the C() I() etc..

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:185]: Agreed! Great Idea.

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:319]: ok, will change these and test.

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:354]: good catch fixing this.

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:375]: They should always be present when a user already exists as homed creates this section, but I do like this idea for safety. will add `.pop` to rest.

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:384]: no, will clean this up. I think single quotes with json are a must as I ran into issues with double quotes.

Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:385]: done.

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py]: done

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:385]: done.

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py]: I changed the logic, let me know if this seems better.

The reason I have all the  'self.result['changed'] = True'  is homed will overwrite any change to user record when using `--identity` so these checks ensure that what the user has configured doesn't get edited/cleared unless they are calling to change it with one of the module options. Let me know if this makes sense or I can try and explain further.

File: tests/integration/targets/homectl/tasks/main.yml
Comment (felixfontein) [inline tests/integration/targets/homectl/tasks/main.yml:1]: You need to add a file tests/integration/targets/homectl/aliases with content `unsupported` to mark that this won't run in CI.

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:343]: I would change the name of this function as it does **not** do the change, but only prepares it.
```suggestion
    def prepare_modify_user_command(self):
```

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:32]: I think the following is more precise:
```suggestion
            - This is required for I(state=present), but will B(only be used) on creation.
```
Is this correct?

File: tests/integration/targets/homectl/tasks/main.yml
Comment (jameslivulpi) [reply tests/integration/targets/homectl/tasks/main.yml:1]: I think these tests can run on CI, there is a condition at bottom of test of what it should be run on:

```
  # homectl was first introduced in systemd 245 so check version  >= 245
  # As of Ubuntu 21 homed does not come with it.
  when: (systemd_version.stdout | regex_search('[0-9][0-9][0-9]') | int >= 245) and ansible_distribution != 'Ubuntu' 
```

Still needs the `unsupported` ?

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:343]: agreed. I will change.

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:32]: Yes, even know when modifying a user we still need the `password` argument as homed requires this to edit the user.

Comment (felixfontein) [reply plugins/modules/system/homectl.py:32]: Does it allow to change the password when editing the user? I.e. what happens if you provide a different password than on the previous run (for the same user)?

File: tests/integration/targets/homectl/tasks/main.yml
Comment (felixfontein) [reply tests/integration/targets/homectl/tasks/main.yml:1]: You can try to make it run in CI. Anyway, `aliases` **must exist** and have the correct entries, otherwise ansible-test ignores this target directory. How about:
```
shippable/posix/group1
skip/aix
skip/freebsd
skip/osx
skip/macos
```
That will filter out the obvious non-systemd systems.

File: tests/integration/targets/homectl/tasks/main.yml
Comment (felixfontein) [inline tests/integration/targets/homectl/tasks/main.yml:162]: Better explicitly test whether homed is available instead of checking `ansible_distribution` (also we don't have Uubuntu 21.x in CI). Maybe run `homectl --version` or `systemctl status systemd-homed` as well and see whether that succeeded?

File: plugins/modules/system/homectl.py
Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:32]: I do plan to add this ( changing current users password) but as of now no.

If you provide a different password it should fail, I just found a bug. I need to add a method to check on users password is correct if its an existing user or else the homectl commands could hang waiting on input.

Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:32]: Next commit will output:
```
fatal: [testhost]: FAILED! => {"changed": false, "msg": "User exists but password is incorrect!", "state": "present"}
```

If the user exists but password is incorrect, I plan to add another field at some point to change password but this will require the user to give the "known" password and the new password to change.

File: tests/integration/targets/homectl/tasks/main.yml
Comment (jameslivulpi) [reply tests/integration/targets/homectl/tasks/main.yml:1]: perfect.

File: tests/integration/targets/homectl/tasks/main.yml
Comment (jameslivulpi) [reply tests/integration/targets/homectl/tasks/main.yml:162]: great added in next commit.

**Comment by felixfontein** [comment]
BTW, I checked the CI logs, it only seems to run in the Fedora 35 instance (homectl is missing in the other ones), but at least it runs somewhere!
https://dev.azure.com/ansible/community.general/_build/results?buildId=34301&view=logs&j=ad47e826-3d51-5052-c5d1-abf94a7b75a5&t=a7141e3f-34f2-5aff-b04a-ec77c032fc27

File: plugins/modules/system/homectl.py
Comment (felixfontein) [reply plugins/modules/system/homectl.py:127]: This still needs to be changed :)

File: plugins/modules/system/homectl.py
Comment (felixfontein) [reply plugins/modules/system/homectl.py:133]: Same here.

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:164]:
```suggestion
            - The keys should follow the same format as the lines in a traditional C(~/.ssh/authorized_key) file.
```

File: plugins/modules/system/homectl.py
Comment (felixfontein) [inline plugins/modules/system/homectl.py:60]: I think `with` makes more sense here (but then, I'm not a native English speaker, so it could also be I'm wrong):
```suggestion
            - This can also be used to add a comment to maintain compatability with C(useradd).
```

Comment (jameslivulpi) [reply plugins/modules/system/homectl.py:60]: Agreed !

**Comment by jameslivulpi** [comment]
> BTW, I checked the CI logs, it only seems to run in the Fedora 35 instance (homectl is missing in the other ones), but at least it runs somewhere! https://dev.azure.com/ansible/community.general/_build/results?buildId=34301&view=logs&j=ad47e826-3d51-5052-c5d1-abf94a7b75a5&t=a7141e3f-34f2-5aff-b04a-ec77c032fc27

Great!

**Review by felixfontein — APPROVED** [review]
How do you want to proceed with this one? It's still marked as "WIP".

**Comment by jameslivulpi** [comment]
Removed WIP tag. Do you think this is ready to be merged?

**Comment by felixfontein** [comment]
It looks polished enough to me for a first version. If there isn't something else you want to extend/change before a first version, we can merge this IMO.

If you or anyone else won't complain, I'll merge by the end of this week :)

**Comment by jameslivulpi** [comment]
Awesome! thanks very much for all the code reviews you did.
You made contributing here very nice! Keep up the good work!!

**Comment by felixfontein** [comment]
@jameslivulpi thanks a lot for your contribution! And you're welcome :)

**Comment by jameslivulpi** [comment]
@felixfontein one last question, is there any openings to become a maintainer for this project?

**Comment by felixfontein** [comment]
@jameslivulpi thank you very much for your interest! We're always looking for people who try helping maintain this all. By contributing this module, you already got the role of module maintainer for this module; if in the future an issue or PR is filed against it, the bot will automatically ping you. But of course you can also do work on other modules and plugins :) Feel free to look at other issues and PRs (or even subscribe to all notifications if you're adventurous ;) ) and see whether there is something you can help with. Sometimes just pointing out something that looks odd can be helpful (especially in issues, a lot of bug reports are not bugs but actually user questions or feature requests disguised as bug reports), or trying to find a reproducer if there is none good included in the issue, or helping test a PR, or even helping to review one.

We regularly reach out to folks who are active over multiple topics for a prolonged time and ask them if they are interested in more privileges, and eventually commit rights. (See https://github.com/ansible-collections/community.general/blob/main/commit-rights.md some writeup on this.) I'm also CC'ing @Andersson007 since I think he wrote more information on this somewhere, which I cannot find right now :)

**Comment by jameslivulpi** [comment]
Thank you Felix, I will try to contribute to issues and reviews as much as I can :-)

**Pull request merged**
